### PR TITLE
[VecOps] Add RVec micro benchmarks

### PR DESCRIPTION
**Pull request by stwunsch**
What do you think? The output on my system looks like follows:

```
9: ------------------------------------------------------------------------
9: Benchmark                              Time             CPU   Iterations
9: ------------------------------------------------------------------------
9: RVecAssign<float>/1                 2.66 ns         2.66 ns    271731449
9: RVecAssign<float>/2                 2.60 ns         2.60 ns    270125305
9: RVecAssign<float>/4                 2.83 ns         2.83 ns    247486718
9: RVecAssign<float>/8                 2.64 ns         2.64 ns    265632530
9: RVecAssign<float>/16                2.64 ns         2.64 ns    265616220
9: RVecAssign<float>/32                3.98 ns         3.98 ns    176340287
9: RVecAssign<float>/64                4.90 ns         4.89 ns    143756388
9: RVecAdd<float>/1                    19.2 ns         19.2 ns     36496547
9: RVecAdd<float>/2                    19.6 ns         19.6 ns     35904417
9: RVecAdd<float>/4                    20.6 ns         20.6 ns     34877230
9: RVecAdd<float>/8                    20.7 ns         20.7 ns     33942916
9: RVecAdd<float>/16                   21.9 ns         21.9 ns     32093654
9: RVecAdd<float>/32                   23.5 ns         23.5 ns     29782320
9: RVecAdd<float>/64                   28.3 ns         28.3 ns     24869119
9: RVecSquareAddSqrt<float>/1          77.5 ns         77.4 ns      9031170
9: RVecSquareAddSqrt<float>/2          92.2 ns         92.1 ns      7589901
9: RVecSquareAddSqrt<float>/4           115 ns          115 ns      6089363
9: RVecSquareAddSqrt<float>/8           165 ns          165 ns      4235401
9: RVecSquareAddSqrt<float>/16          272 ns          272 ns      2567060
9: RVecSquareAddSqrt<float>/32          479 ns          478 ns      1446756
9: RVecSquareAddSqrt<float>/64          890 ns          889 ns       786075
9: RVecDeltaR<float>/1                  141 ns          141 ns      5163104
9: RVecDeltaR<float>/2                  146 ns          145 ns      4842268
9: RVecDeltaR<float>/4                  158 ns          158 ns      4418587
9: RVecDeltaR<float>/8                  179 ns          178 ns      3938653
9: RVecDeltaR<float>/16                 233 ns          233 ns      3024173
9: RVecDeltaR<float>/32                 322 ns          321 ns      2176568
9: RVecDeltaR<float>/64                 501 ns          500 ns      1388788
9: RVecInvariantMass<float>/1          15.2 ns         15.2 ns     45623921
9: RVecInvariantMass<float>/2          25.1 ns         25.1 ns     27850061
9: RVecInvariantMass<float>/4          43.9 ns         43.9 ns     15824374
9: RVecInvariantMass<float>/8          81.4 ns         81.4 ns      8513980
9: RVecInvariantMass<float>/16          153 ns          153 ns      4572423
9: RVecInvariantMass<float>/32          299 ns          299 ns      2346282
9: RVecInvariantMass<float>/64          583 ns          583 ns      1194280
9: RVecInvariantMasses<float>/1        38.8 ns         38.7 ns     17867903
9: RVecInvariantMasses<float>/2        58.8 ns         58.8 ns     11671040
9: RVecInvariantMasses<float>/4         101 ns          100 ns      6990254
9: RVecInvariantMasses<float>/8         188 ns          187 ns      3731681
9: RVecInvariantMasses<float>/16        351 ns          351 ns      1996413
9: RVecInvariantMasses<float>/32        683 ns          682 ns      1013530
9: RVecInvariantMasses<float>/64       1368 ns         1366 ns       518215
9: RVecMaskingSimple<float>/1          28.5 ns         28.5 ns     25180291
9: RVecMaskingSimple<float>/2          29.6 ns         29.5 ns     23570766
9: RVecMaskingSimple<float>/4          32.7 ns         32.7 ns     21282888
9: RVecMaskingSimple<float>/8          40.2 ns         40.1 ns     17255973
9: RVecMaskingSimple<float>/16         54.0 ns         54.0 ns     12662199
9: RVecMaskingSimple<float>/32         84.5 ns         84.5 ns      8211300
9: RVecMaskingSimple<float>/64          146 ns          146 ns      4793308
9: RVecMaskingComplex<float>/1          115 ns          115 ns      6079756
9: RVecMaskingComplex<float>/2          121 ns          121 ns      5799021
9: RVecMaskingComplex<float>/4          133 ns          133 ns      5228572
9: RVecMaskingComplex<float>/8          153 ns          153 ns      4587173
9: RVecMaskingComplex<float>/16         197 ns          197 ns      3541794
9: RVecMaskingComplex<float>/32         278 ns          278 ns      2507866
9: RVecMaskingComplex<float>/64         454 ns          454 ns      1538917
1/1 Test #9: rootbench-VecOpsRVec .............   Passed   52.71 sec
```

### CODE REVIEW

**Comment by stwunsch** [comment]
Here a flamegraph showing the search for `(malloc|free)`, which matches 21%:
Screenshot from 2020-10-02 13-56-15

**Comment by stwunsch** [comment]
Looking at the prime candidates for SBO: The simple addition of RVecs and also the masking benchmarks spent around 40% of their time in `(free|malloc)`. See the screenshots below!
Screenshot from 2020-10-02 14-06-21
Screenshot from 2020-10-02 14-06-13
Screenshot from 2020-10-02 14-06-03

**Comment by vgvassilev** [comment]
Nice. Long time ago I spent some time trying to condense the vectorization benchmark outputs. You can take a look at Cartesian3DBenchmarks and alike and maybe you like it.

**Comment by stwunsch** [comment]
> Nice. Long time ago I spent some time trying to condense the vectorization benchmark outputs. You can take a look at Cartesian3DBenchmarks and alike and maybe you like it.

Thanks for the hint! However, for now, the main purpose of these benchmarks is the issue with too many allocations if you use `RVec` expressions like `sqrt(pow(rvec, 2) + pow(rvec, 2))`. But I agree that we should keep an eye on vectorization too.

**Comment by vgvassilev** [comment]
That is an interesting use case. Would that be a more generic way to do what you intend -- https://github.com/google/benchmark/pull/625

**Comment by stwunsch** [comment]
Oh nice! Have never seen this in any documentation? Seems like the code and [this test](https://github.com/google/benchmark/blob/master/test/memory_manager_test.cc) is the documentation?

**Comment by vgvassilev** [comment]
Code and tests are the true documentation for open source ;)

**Comment by stwunsch** [comment]
That's a great idea!

**Review by eguiraud — APPROVED** [review]
LGTM!

File: root/math/vecops/VecOpsRVec.cxx [hunk 1]
```diff
@@ -0,0 +1,139 @@
+#include "ROOT/RVec.hxx"
+#include "benchmark/benchmark.h"
+#include "rootbench/RBConfig.h"
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply root/math/vecops/VecOpsRVec.cxx]: Thanks! Fixed

File: root/math/vecops/VecOpsRVec.cxx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+template <class T>
+static void RVecAssign(benchmark::State &state)
+{
+   const auto size = state.range(0);
+   RVec<T> v1(size);
+   RVec<T> v2(size);
+   for (auto _ : state) {
+      v2 = v1;
+      benchmark::DoNotOptimize(v2);
+   }
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply root/math/vecops/VecOpsRVec.cxx:32]: Yep, you are right. I've now initialized all vectors with (sequential) values to be very sure that things are computed as expected.

**Pull request merged**